ehrapy/tools/cohort_tracking/_cohort_tracker.py

```diff
@@ -0,0 +1,332 @@
+        self._tracked_text: list = []
+        self._tracked_operations: list = []
+
+        # if categorical columns specified, use them
+        # else, follow tableone's logic
+        self.categorical = categorical if categorical is not None else _detect_categorical_columns(df[self.columns])
+        self.track = self._get_column_structure(df)
+
+        self._track_backup = copy.deepcopy(self.track)
+
+    def __call__(
+        self, adata: AnnData, label: str = None, operations_done: str = None, *args: Any, **tableone_kwargs: Any
+    ) -> Any:
+        if isinstance(adata, AnnData):
+            df = adata.obs
+        elif isinstance(adata, pd.DataFrame):
+            df = adata
+        else:
+            raise ValueError("adata must be an AnnData or a DataFrame.")
+
+        _check_columns_exist(df, self.columns)
+
+        # track a small text with each tracking step, for the flowchart
+        track_text = label if label is not None else f"Cohort {self.tracked_steps}"
+        track_text += "\n (n=" + str(adata.n_obs) + ")"
+        self._tracked_text.append(track_text)
+
+        # track a small text with the operations done
+        self._tracked_operations.append(operations_done)
+
+        self._tracked_steps += 1
+
+        t1 = TableOne(df, columns=self.columns, categorical=self.categorical, **tableone_kwargs)
+        # track new stuff
+        self._get_column_dicts(t1)
+
+    def _get_column_structure(self, df):
+        column_structure = {}
+        for column in self.columns:
+            if column in self.categorical:
+                # if e.g. a column containing integers is deemed categorical, coerce it to categorical
+                df[column] = df[column].astype("category")
+                column_structure[column] = {category: [] for category in df[column].cat.categories}
+            else:
+                column_structure[column] = []
+
+        return column_structure
+
+    def _get_column_dicts(self, table_one):
+        for col, value in self.track.items():
+            if isinstance(value, dict):
+                self._get_cat_dicts(table_one, col)
+            else:
+                self._get_num_dicts(table_one, col)
+
+    def _get_cat_dicts(self, table_one, col):
+        for cat in self.track[col].keys():
+            # if tableone does not have the category of this column anymore, set the percentage to 0
+            # for categorized columns (e.g. gender 1.0/0.0), str(cat) helps to avoid considering the category as a float
+            if (col, str(cat)) in table_one.cat_table["Overall"].index:
+                pct = float(table_one.cat_table["Overall"].loc[(col, str(cat))].split("(")[1].split(")")[0])
+            else:
+                pct = 0
+            self.track[col][cat].append(pct)
+
+    def _get_num_dicts(self, table_one, col):
+        summary = table_one.cont_table["Overall"].loc[(col, "")]
+        self.track[col].append(summary)
+
+    def reset(self):
```

Comment:
It makes sense to have it, but maybe people should instead just create new objects.

Reply:
Yeah I was not sure if that's even a good idea to have, the overhead of creating a new object's not that big indeed...